spacy/cli/project/parallel.py

```diff
@@ -0,0 +1,442 @@
+        if (
+            max_parallel_processes is not None
+            and max_parallel_processes < num_concurr_worker_processes
+        ):
+            num_concurr_worker_processes = max_parallel_processes
+        worker_process_iterator = iter(worker_processes)
+        for _ in range(num_concurr_worker_processes):
+            _start_worker_process(next(worker_process_iterator), proc_to_cmd_infos)
+        divider_parallel_descriptor = parallel_descriptor = (
+            "parallel[" + ", ".join(cmd_info.cmd_name for cmd_info in cmd_infos) + "]"
+        )
+        if len(divider_parallel_descriptor) > MAX_WIDTH_DIVIDER:
+            divider_parallel_descriptor = (
+                divider_parallel_descriptor[: (MAX_WIDTH_DIVIDER - 3)] + "..."
+            )
+        msg.divider(divider_parallel_descriptor)
+        if not DISPLAY_STATUS_TABLE and len(parallel_descriptor) > MAX_WIDTH_DIVIDER:
+            # reprint the descriptor if it was too long and had to be cut short
+            print(parallel_descriptor)
+
+        msg.info(
+            "Temporary logs are being written to "
+            + os.sep.join((str(current_dir), PARALLEL_LOGGING_DIR_NAME))
+        )
+        first = True
+        while any(
+            cmd_info.state.name in ("pending", "starting", "running")
+            for cmd_info in cmd_infos
+        ):
+            try:
+                mess: Dict[str, Union[str, int]] = parallel_group_status_queue.get(
+                    timeout=PARALLEL_GROUP_STATUS_INTERVAL * 20
+                )
+            except Exception:
+                # No more messages are being received: the whole group has hung
+                for other_cmd_info in (
+                    c for c in cmd_infos if c.state.name in ("starting", "running")
+                ):
+                    other_cmd_info.change_state("hung")
+                for other_cmd_info in (
+                    c for c in cmd_infos if c.state.name == "pending"
+                ):
+                    other_cmd_info.change_state("cancelled")
+                break
+            cmd_info = cmd_infos[cast(int, mess["cmd_index"])]
+            if mess["type"] in ("started", "keepalive"):
+                cmd_info.last_keepalive_time = int(time())
+            for other_cmd_info in (
+                c for c in cmd_infos if c.state.name in ("starting", "running")
+            ):
+                if (
+                    other_cmd_info.last_keepalive_time is not None
+                    and time() - other_cmd_info.last_keepalive_time
+                    > PARALLEL_GROUP_STATUS_INTERVAL * 20
+                ):
+                    # a specific command has hung
+                    other_cmd_info.change_state("hung")
+            if mess["type"] == "started":
+                cmd_info.change_state("running")
+                cmd_info.running_os_cmd_index = cast(int, mess["os_cmd_index"])
+                cmd_info.pid = cast(int, mess["pid"])
```

Comment:
IMO the `cast` isn't necessary here. I don't see any wrong type hints without it, at least.

Reply:
Strange, my `mypy` complains. The problem is that the dictionary entries in the message have the type `Union[str, int]` and have to be cast to whichever type the target variable has.